### PR TITLE
use metacpan download_url endpoint

### PR DESCRIPTION
**Pull request by haarg**

### CODE REVIEW

**Comment by miyagawa** [comment]
Is this the same as #520?

**Comment by haarg** [comment]
This is against the old cpanm codebase.  #520 is against Menlo.

**Comment by skaji** [comment]
* MetaCPAN is https
* `has_working_lwp` method only checks scheme for $mirrors  
https://github.com/miyagawa/cpanminus/blob/devel/lib/App/cpanminus/script.pm#L2917

So, if LWP is installed, but not LWP::Protocol::https, then cpanm will fail to access MetaCPAN.

**Comment by miyagawa** [comment]
Good point. Is it not an issue in Menlo because it uses HTTP::Tiny?

**Comment by haarg** [comment]
While this does change the URL used from http to https, api.metacpan.org always redirects to https.  So the issue with LWP exists even without this PR.

**Comment by skaji** [comment]
@miyagawa 
> Is it not an issue in Menlo because it uses HTTP::Tiny?

Yes, it is.

@haarg 
> api.metacpan.org always redirects to https. 

Oops. I think we used to be able to access api.metacpan.org via http.
Did metacpan-team change the scheme of api.metacpan.org during meta::hack?

File: lib/App/cpanminus/script.pm [hunk 1]
```diff
@@ -553,126 +523,31 @@ sub numify_ver {
...
     if ($dist_meta && $dist_meta->{download_url}) {
         (my $distfile = $dist_meta->{download_url}) =~ s!.+/authors/id/!!;
         local $self->{mirrors} = $self->{mirrors};
-        if ($dist_meta->{status} eq 'backpan') {
-            $self->{mirrors} = [ 'http://backpan.perl.org' ];
-        } elsif ($dist_meta->{stat}{mtime} > time()-24*60*60) {
-            $self->{mirrors} = [ 'http://cpan.metacpan.org' ];
-        }
-        return $self->cpan_module($module, $distfile, $module_version);
+        $self->{mirrors} = [ 'https://cpan.metacpan.org' ];
```
Comment (skaji) [inline lib/App/cpanminus/script.pm]: I don't have a strong opinion, but do we always need to choose cpan.metacpan.org as a mirror here?

Comment (haarg) [inline lib/App/cpanminus/script.pm]: Part of the reason for this is that cpan.metacpan.org is a backpan.  This routine is mainly used for complex version specifications, which will often need a backpan, and by --metacpan, which it makes sense to switch the mirror for.  I also don't really see a downside.  It could be refactored to handle this somewhat differently, but I'd rather that be addressed in a separate PR.

Comment (skaji) [inline lib/App/cpanminus/script.pm]: I see.

File: lib/App/cpanminus/script.pm [hunk 2]
```diff
@@ -2923,7 +2798,7 @@ sub init_tools {
     }
 
     # use --no-lwp if they have a broken LWP, to upgrade LWP
-    if ($self->{try_lwp} && $self->has_working_lwp($self->{mirrors})) {
+    if ($self->{try_lwp} && $self->has_working_lwp($self->{mirrors}, $self->{metacpan})) {
```
Comment (skaji) [inline lib/App/cpanminus/script.pm]: $self->{metacpan} is not a flag for whether cpanm accesses MetaCPAN or not.

**Comment by haarg** [comment]
I was actually wrong about api.metacpan.org always redirecting to https.  That was being done by my browser.

**Comment by haarg** [comment]
It would be possible for metacpan to allow non-https access to the API.  Honestly, I think it's rather questionable for cpanm to use non-https at all, at least without explicit opt-in.  But if that would make this changeover easier to accept, we can make something work.

File: lib/App/cpanminus/script.pm (hunk 1 above)
Comment (ranguard) [inline lib/App/cpanminus/script.pm]: FYI: backpan.metacpan.org exists (it's just an alias for cpan.metacpan.org - but makes the distinction clear)

**Comment by miyagawa** [comment]
been fuzzy on the details of this - i merged the menlo version of the PR today, what should be done on this?

**Comment by skaji** [comment]
It is no wonder that MetaCPAN serves via https.
So how about checking LWP::Protocol::https always?
```diff
diff --git a/lib/App/cpanminus/script.pm b/lib/App/cpanminus/script.pm
index 38b6bcc..ae63240 100644
--- a/lib/App/cpanminus/script.pm
+++ b/lib/App/cpanminus/script.pm
@@ -2927,12 +2927,11 @@ sub file_mirror {
 }
 
 sub has_working_lwp {
-    my($self, $mirrors) = @_;
-    my $https = grep /^https:/, @$mirrors;
+    my $self = shift;
     eval {
         require LWP::UserAgent; # no fatpack
         LWP::UserAgent->VERSION(5.802);
-        require LWP::Protocol::https if $https; # no fatpack
+        require LWP::Protocol::https; # no fatpack
         1;
     };
 }
@@ -2947,7 +2946,7 @@ sub init_tools {
     }
 
     # use --no-lwp if they have a broken LWP, to upgrade LWP
-    if ($self->{try_lwp} && $self->has_working_lwp($self->{mirrors})) {
+    if ($self->{try_lwp} && $self->has_working_lwp) {
         $self->chat("You have LWP $LWP::VERSION\n");
         my $ua = sub {
             LWP::UserAgent->new(

```

**Comment by haarg** [comment]
I didn't plan on pursuing this any further.  Menlo will be using the new API, and we now have a shim in place so cpanm 1.7 will use the v1 API behind the scenes.  So cpanm is no longer holding us up from dropping the MetaCPAN v0 API.
